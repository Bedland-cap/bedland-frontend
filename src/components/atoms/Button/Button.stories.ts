import type { Meta, StoryObj } from '@storybook/react';

import { Button } from './Button';

// docs: https://storybook.js.org/docs/7.0/react/writing-stories/introduction
const meta = {
  title: 'atoms/Button',
  component: Button,
  tags: ['autodocs'],
  argTypes: {
    backgroundColor: { control: 'color' },
  },
} satisfies Meta<typeof Button>;

export default meta;
type Story = StoryObj<typeof meta>;

export const Primary: Story = {
  args: {
    primary: true,
    label: 'Press me',
  },
};

export const Secondary: Story = {
  args: {
    label: 'Press me',
  },
};

Secondary.parameters = {
  backgrounds: { default: 'light' },
};

export const Large: Story = {
  args: {
    size: 'large',
    label: 'Press me',
  },
};

Large.parameters = {
  backgrounds: { default: 'light' },
};

export const Small: Story = {
  args: {
    size: 'small',
    label: 'Press me',
  },
};

Small.parameters = {
  backgrounds: { default: 'light' },
};
