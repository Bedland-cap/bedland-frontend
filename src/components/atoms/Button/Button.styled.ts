import styled from 'styled-components';

const Btn = styled.button<{
  backgroundColor: string;
  textColor: string;
  variant: 'primary' | 'secondary';
}>`
  align-items: center;
  background-color: ${(props) => props.backgroundColor};
  border: 1px solid black;
  border-radius: 0.5rem;
  color: ${(props) => props.textColor};
  display: flex;
  height: 2.5rem;
  justify-content: center;
  padding: 0.625rem 1rem 0.625rem 1rem;
  width: 19.875rem;
  &:disabled,
  &:hover,
  &:active,
  &:focus {
    outline: none;
  }
`;
export default Btn;
