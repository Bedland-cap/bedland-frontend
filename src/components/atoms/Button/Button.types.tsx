import { ButtonHTMLAttributes, CSSProperties } from 'react';
import { ColorNames, HexColor } from 'theme/theme.types';
import { IconName } from '../Icon/icon.types';

export const ButtonVariants = ['primary', 'text', 'ghost', 'icon', 'notificationItem'] as const;
type ButtonVariantsType = {
  [key in (typeof ButtonVariants)[number]]: key;
};
export const buttonColors = ['orange', 'blue'] as const;
export type ColorVariant = (typeof buttonColors)[number];

export const BUTTON_VARIANTS: ButtonVariantsType = {
  primary: 'primary',
  text: 'text',
  ghost: 'ghost',
  icon: 'icon',
  notificationItem: 'notificationItem',
} as const;

export type ButtonVariantProps = (typeof ButtonVariants)[number];

export type WrapperProps = {
  wrapper?: true;
  width?: number;
  height?: number;
  marginTop?: number;
  style?: CSSProperties;
};
type WrapperPropsNever = {
  wrapper?: never;
  width?: never;
  height?: never;
  marginTop?: never;
  style?: never;
};

export type IconProps = { icon?: IconName | null; iconSize?: number; iconColor?: ColorNames };

type BasicProps<T extends ButtonVariantProps> = { variant: T; color?: ColorVariant };
type BasicPropsNever = { variant?: never; color?: never };
export type ButtonWrapperProps = BasicPropsNever & WrapperProps & IconProps;

export type PrimaryButtonProps = BasicProps<typeof BUTTON_VARIANTS.primary> &
  WrapperPropsNever &
  IconProps;

export type TextButtonProps = BasicProps<typeof BUTTON_VARIANTS.text> &
  WrapperPropsNever &
  IconProps;

export type GhostButtonProps = BasicProps<typeof BUTTON_VARIANTS.ghost> &
  WrapperPropsNever &
  IconProps;

export type IconButtonProps = BasicProps<typeof BUTTON_VARIANTS.icon> &
  WrapperPropsNever &
  IconProps;

export type NotificationItemButtonProps = BasicProps<typeof BUTTON_VARIANTS.notificationItem> &
  WrapperPropsNever &
  IconProps;

export type ButtonProps = (
  | PrimaryButtonProps
  | TextButtonProps
  | GhostButtonProps
  | IconButtonProps
  | ButtonWrapperProps
  | NotificationItemButtonProps
) &
  ButtonHTMLAttributes<HTMLButtonElement>;

export type StylingProps = {
  styling: {
    bckgColor: HexColor | 'transparent';
    borderColor: HexColor | 'transparent';
    color: HexColor;
    hoverBckgColor: HexColor | 'transparent';
    hoverBorderColor: HexColor | 'transparent';
    paddingTB?: number;
    pressedBckgColor?: HexColor | 'transparent';
    pressedBorderColor?: HexColor | 'transparent';
  };
};

export type ColorVariants = {
  [color in ColorVariant | 'noStyleColor']: {
    [variant in ButtonVariantProps]?: StylingProps['styling'];
  };
};

export type StyledWrapperProps = {
  styling: WrapperProps;
};

export type ButtonWrapperValueType<T> = {
  ButtonWrapper: T;
} & StyledWrapperProps;

export type ButtonsStyleValueType<T> = {
  StyledButton: T;
} & StylingProps;

export type ButtonsStyles<T> = {
  [key2 in ButtonVariantProps]: (color: ColorVariant) => ButtonsStyleValueType<T>;
};

export type ButtonsStyleType<T> = {
  [key in keyof ButtonVariantsType]: ButtonsStyleValueType<T>;
};

export const IsIconNullGuard = (icon: null | IconName): icon is null => icon === null;
