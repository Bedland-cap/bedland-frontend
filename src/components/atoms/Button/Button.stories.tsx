import type { Meta, StoryObj } from '@storybook/react';
import GlobalStyle from 'theme/globalStyles';
import ThemeProvider from 'theme/ThemeContext';
import Button from './Button';

const meta = {
  title: 'Atoms/Button',
  component: Button,
  tags: ['autodocs'],
  decorators: [
    (Story) => (
      <ThemeProvider>
        <GlobalStyle />
        {Story()}
      </ThemeProvider>
    ),
  ],
} satisfies Meta<typeof Button>;

export default meta;
type Story = StoryObj<typeof meta>;

export const Default: Story = {
  args: {
    backgroundColor: '#ffffff',
    children: 'Login',
    type: 'button',
    textColor: '#000000',
    variant: 'primary',
  },
};
