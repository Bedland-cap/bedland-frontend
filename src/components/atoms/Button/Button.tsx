import { PropsWithChildren } from 'react';
import * as Styled from './Button.styled';

type ButtonProps = {
  backgroundColor: string;
  type: 'submit' | 'reset' | 'button' | undefined;
  textColor: string;
  variant: 'primary' | 'secondary';
};

const Button = ({
  backgroundColor,
  textColor,
  children,
  type = 'button',
  variant = 'primary',
}: PropsWithChildren<ButtonProps>) => (
  <Styled.Btn type={type} backgroundColor={backgroundColor} textColor={textColor} variant={variant}>
    {children}
  </Styled.Btn>
);

export default Button;
