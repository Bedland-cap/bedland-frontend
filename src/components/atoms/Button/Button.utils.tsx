import initialTheme from 'theme/initialTheme';
import { colorNames, ColorNames, HexColor } from 'theme/theme.types';
import {
  ButtonsStyleValueType,
  ButtonVariantProps,
  WrapperProps,
  ButtonWrapperValueType,
  BUTTON_VARIANTS,
  ButtonsStyles,
  ColorVariants,
  ColorVariant,
  StylingProps,
} from './Button.types';
import * as Styled from './Button.styled';

export const COLOR_VARIANTS: ColorVariants = {
  orange: {
    primary: {
      bckgColor: initialTheme.palette.orangeDark,
      borderColor: initialTheme.palette.orangeDark,
      color: initialTheme.palette.light,
      hoverBckgColor: initialTheme.palette.orangeLight,
      hoverBorderColor: initialTheme.palette.orangeLight,
    },
    text: {
      bckgColor: 'transparent',
      borderColor: 'transparent',
      color: initialTheme.palette.orangeLight,
      hoverBckgColor: `${initialTheme.palette.orangeLight}0D`,
      hoverBorderColor: 'transparent',
      pressedBckgColor: `${initialTheme.palette.orangeLight}26`,
      pressedBorderColor: 'transparent',
    },
    ghost: {
      bckgColor: 'transparent',
      borderColor: initialTheme.palette.orangeDark,
      color: initialTheme.palette.orangeDark,
      hoverBckgColor: `${initialTheme.palette.orangeDark}1A`,
      hoverBorderColor: initialTheme.palette.orangeDark,
    },
  },
  blue: {
    primary: {
      bckgColor: initialTheme.palette.blueDark,
      borderColor: initialTheme.palette.blueDark,
      color: initialTheme.palette.light,
      hoverBckgColor: initialTheme.palette.blueLight,
      hoverBorderColor: initialTheme.palette.blueLight,
    },
    text: {
      bckgColor: 'transparent',
      borderColor: 'transparent',
      color: initialTheme.palette.blueDark,
      hoverBckgColor: `${initialTheme.palette.blueDark}0D`,
      hoverBorderColor: 'transparent',
      pressedBckgColor: `${initialTheme.palette.blueDark}26`,
      pressedBorderColor: 'transparent',
    },
    ghost: {
      bckgColor: 'transparent',
      borderColor: initialTheme.palette.blueDark,
      color: initialTheme.palette.blueDark,
      hoverBckgColor: `${initialTheme.palette.blueDark}1A`,
      hoverBorderColor: initialTheme.palette.blueDark,
    },
  },
  noStyleColor: {
    primary: {
      bckgColor: 'transparent',
      borderColor: 'transparent',
      color: initialTheme.palette.blueDark,
      hoverBckgColor: `${initialTheme.palette.orangeLight}0D`,
      hoverBorderColor: 'transparent',
      pressedBckgColor: `${initialTheme.palette.orangeLight}26`,
      pressedBorderColor: 'transparent',
    },
  },
};

export const BUTTONS_STYLES: ButtonsStyles<typeof Styled.PrimaryButton> = {
  primary: (color) => ({
    StyledButton: Styled.PrimaryButton,
    styling: COLOR_VARIANTS[color][BUTTON_VARIANTS.primary] as StylingProps['styling'],
  }),
  text: (color) => ({
    StyledButton: Styled.TextButton,
    styling: COLOR_VARIANTS[color][BUTTON_VARIANTS.text] as StylingProps['styling'],
  }),
  ghost: (color) => ({
    StyledButton: Styled.GhostButton,
    styling: COLOR_VARIANTS[color][BUTTON_VARIANTS.ghost] as StylingProps['styling'],
  }),
  icon: () => ({
    StyledButton: Styled.IconButton,
    styling: COLOR_VARIANTS.noStyleColor[BUTTON_VARIANTS.primary] as StylingProps['styling'],
  }),
  notificationItem: (color) => ({
    StyledButton: Styled.GhostButton,
    styling: {
      ...(COLOR_VARIANTS[color][BUTTON_VARIANTS.ghost] as StylingProps['styling']),
      paddingTB: 0.3125,
    },
  }),
};

export const getStyledButton = (
  variant: ButtonVariantProps,
  color: ColorVariant,
): ButtonsStyleValueType<typeof Styled.PrimaryButton> => BUTTONS_STYLES[variant](color);

export const getWrapper = ({
  width,
  height,
  marginTop,
}: WrapperProps): ButtonWrapperValueType<typeof Styled.ButtonWrapper> => ({
  ButtonWrapper: Styled.ButtonWrapper,
  styling: {
    width,
    height,
    marginTop,
  },
});

export const getIconColor = (color: HexColor): ColorNames =>
  colorNames.find((key) => initialTheme.palette[key] === color) as ColorNames;
