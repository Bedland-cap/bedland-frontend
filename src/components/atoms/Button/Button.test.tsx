import { cleanup } from '@testing-library/react';
import renderer from 'react-test-renderer';
import ThemeProvider from 'theme/ThemeContext';
import Button from './Button';

describe('Button', () => {
  afterEach(cleanup);
  it('renders correctly', () => {
    const snapshot = renderer
      .create(
        <ThemeProvider>
          <Button backgroundColor='#ffffff' type='button' variant='primary' textColor='#ffffff'>
            login
          </Button>
        </ThemeProvider>,
      )
      .toJSON();
    expect(snapshot).toMatchSnapshot();
  });
});
