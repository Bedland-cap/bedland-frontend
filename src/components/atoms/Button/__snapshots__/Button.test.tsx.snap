// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button renders correctly 1`] = `
.c0 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background-color: #ffffff;
  border: 1px solid black;
  border-radius: 0.5rem;
  color: #ffffff;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  height: 2.5rem;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  padding: 0.625rem 1rem 0.625rem 1rem;
  width: 19.875rem;
}

.c0:disabled,
.c0:hover,
.c0:active,
.c0:focus {
  outline: none;
}

<button
  className="c0"
  type="button"
>
  login
</button>
`;
